stmocli/stmo.py

```diff
@@ -86,3 +86,27 @@ def fork_query(self, query_id, new_query_file_name):
         result = self._redash.fork_query(query_id)
         fork = self.track_query(result["id"], new_query_file_name)
         return fork
+
+    def get_results(self, query_id):
+        """Pull the existing dataset from a query using the query id
+        Doesn't have to be tracked - we will pull the query syntax from stmo
+        and use that.
+
+        Args:
+            query_id: redash id of the query
+
+        Returns:
+            1. The name of the query, to be used for csv filenames if desired
+            2. List of dicts, one dict per row of results. Dicts are keyed by
+            column labels of the returned data.
+
+        Raises:
+            RedashClientException if query id not found on redash
+        """
+        query = self.get_query(query_id)
+        sql_query = query["query"]
+        query_info = QueryInfo.from_dict(query)
+        data_source_id = query_info.data_source_id
+        query_name = query_info.name
+        results = self._redash.get_query_results(sql_query, data_source_id)
```

Comment:
I think `RedashClient.get_query_results` is doing the wrong thing here.

These are the maps between Redash's URLs and handlers. Note the Redash API has an endpoint for getting a CSV for a query number: https://github.com/getredash/redash/blob/35bb558f442792f9fd92793fda45763a3d134fc5/redash/handlers/api.py#L92-L97

But RedashClient isn't using that -- instead, RedashClient is calling the /query_results parameter with no additional URI parameters and posting in a blob of SQL instead of a query number: https://github.com/mozilla/redash_client/blob/master/redash_client/client.py#L147

Which maps on to the QueryResultListResource, which seems to execute the query on each call?: https://github.com/getredash/redash/blob/1e6d6bbeb3ad50f531fb017b8abec7e7a7ffb79c/redash/handlers/query_results.py#L99-L129

I think the code path we want is through QueryResultResource, which uses cached data: https://github.com/getredash/redash/blob/1e6d6bbeb3ad50f531fb017b8abec7e7a7ffb79c/redash/handlers/query_results.py#L159

To do that, we want to fetch a URL that looks like `/api/queries/<query_id>/results.<filetype>`.

So I think we should either implement our own method for reaching that URL by abusing `_redash._make_request` again, or else send redash_client a PR for a new method. (I have been avoiding the latter option, perhaps irresponsibly)

Does that seem right to you?

Reply:
You're right that `get_query_results` *should* be executing sql code, rather than fetching results based on a query number. But weirdly, I think there may be some kind of caching going on on the redash server... for example when I test this implementation with query number `50305` which takes an hour to run, i get the cached results back instantly. And I haven't run that query in 8 months. Either way though, I should probably change this to not use `get_query_results`

Reply:
Oh, there's a bonus documentation bug in redash.

What `QueryResultListResource` _says_ about the max_age parameter, which RedashClient omits:

> :qparam number max_age: If query results less than `max_age` seconds old are available, return them, otherwise execute the query; **if omitted, always execute**

This is what would happen if it defaulted max_age to 0. But what it actually _does_ is [default max_age to -1](https://github.com/getredash/redash/blob/1e6d6bbeb3ad50f531fb017b8abec7e7a7ffb79c/redash/handlers/query_results.py#L114), which means [any cached result of any age](https://github.com/getredash/redash/blob/1e6d6bbeb3ad50f531fb017b8abec7e7a7ffb79c/redash/models.py#L754).

So it definitely tries to return _a _ cached result.

(In the edge case where you have two redash queries with the same query string that were refreshed at different times, I wonder if refreshing the results for one of the redash query IDs refreshes the results for the other query ID...)

Reply:
Upstream PR, for reference https://github.com/getredash/redash/pull/2815